feat(perl)!: replace parser

rabbiveesh opened the pull request:
This replaces the perl parser from ganezdragon with an actively maintained + considerably
more correct version. There are some features that we're still working on, but by and
large this has already proven very useful for me in my day to day.

One of the things that we support is coloring scalars, arrays and hashes differently. Is
there a way to expose that in a simple way to users? We're doing it by making a
`@variable.scalar`, `@variable.array` and `@variable.hash`.

clason commented:
Is this the one @leonerd has been working on?

clason commented on queries/perl/injections.scm:
```diff
@@ -1 +1,2 @@
-(comments) @comment
```
We globally have that comment injection; any reason to remove it?

leonerd replied:
I'm not sure this is particularly a concsious decision to remove it, so much as just that our copy started off empty, and we added `(pod)`, and didn't add this one that you have.

That said, I don't think we have a `(comments)` definition, do we? Just a single-line `(comment)`.

clason replied:
Well, whatever comment node you have should inject the `comment` parser.

leonerd commented:
> Is this the one @leonerd has been working on?

It is, yes

clason commented:
So, first order of business:

1. we use a different query precedence than upstream (for now; they'll change eventually): latest match wins
2. we use different capture names, see https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights -- any other (toplevel) captures are illegal and trip CI.

ObserverOfTime commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,89 @@
+  "or" "and"
+  "eq" "ne" "cmp" "lt" "le" "ge" "gt"
+  "isa"
+] @operator
```
```suggestion
] @keyword.operator
```

ObserverOfTime commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,89 @@
+(interpolated_string_literal) @string
+(quoted_word_list) @string
+(command_string) @string
+[(heredoc_token) (command_heredoc_token)] @string.special
+(heredoc_content) @string
+(heredoc_end) @string.special
```
We generally use `@label` for heredoc tokens.
```suggestion
[(heredoc_token) (command_heredoc_token)] @label
(heredoc_content) @string
(heredoc_end) @label
```

ObserverOfTime commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,89 @@
 
-; version sortof counts as a kind of multipart integer
-(version) @constant
+[(quoted_regexp) (match_regexp)] @string.regexp
```
```suggestion
[(quoted_regexp) (match_regexp)] @string.regex
```

ObserverOfTime commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,89 @@
-   "exists" "delete" "keys" "values"
-   "each"))
+(subroutine_declaration_statement name: (_) @function)
+(attribute_name) @decorator
```
```suggestion
(attribute_name) @attribute
```

leonerd commented:
> * we use a different query precedence than upstream (for now; they'll change eventually): latest match wins

Ah as a total aside, I was curious about that for my own implementation. I think this seems to be the most common interpretation. Good.

But I'm unclear why you bring it up specifically now - I don't think we rely on that anywhere, do we? I've mostly been using/testing this via nvim anyway, so I believe nvim's interpretation is correct.

> * we use different capture names, see https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights -- any other (toplevel) captures are illegal and trip CI.

Ahyes; probably some adjusting needs doing there then.

ObserverOfTime commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,89 @@
-; Misc keywords
 [
```
Keep the previous captures for these:

```scheme
[ "use" "no" "require" ] @include

[ "if" "elsif" "unless" "else" ] @conditional

[ "while" "until" "for" "foreach" ] @repeat

"return" @keyword.return

"sub" @keyword.function
```

Also, "map" and "grep" could be `@function.builtin`.

ObserverOfTime commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,89 @@
 
-(named_block_statement
-  function_name: (identifier) @function)
+(function_call_expression (function) @function)
+(method_call_expression (method) @function.method)
```
```suggestion
(function_call_expression (function) @function.call)
(method_call_expression (method) @method.call)
```

ObserverOfTime commented on queries/perl/injections.scm:
```diff
@@ -1 +1,2 @@
-(comments) @comment
+; an injections.scm file for nvim-treesitter
+(pod) @pod
```
Can you also add the pod parser (or comment this out till then)?

leonerd replied:
Oh; the POD parser is at https://github.com/tree-sitter-perl/tree-sitter-pod/

rabbiveesh replied:
I'll do this today, I added a release branch there too just now

rabbiveesh replied:
has been added

ObserverOfTime reviewed (changes requested):
Missing highlights for operators and punctuation.

Also missing this for shebangs:
```scheme
((source_file . (comment) @preproc)
  (#lua-match? @preproc "^#!/"))
```

clason commented:
> Ah as a total aside, I was curious about that for my own implementation. I think this seems to be the most common interpretation. Good.

It's the interpretation upstream will switch to, as far as I know.

> But I'm unclear why you bring it up specifically now - I don't think we rely on that anywhere, do we? I've mostly been using/testing this via nvim anyway, so I believe nvim's interpretation is correct.

It's just a very common gotcha when queries are downstreamed, so I thought I'd mention it just to potentially save some time.

lucario387 commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,89 @@
 
-(function_definition
-  name: (identifier) @function)
+(func0op_call_expression function: _ @function.builtin)
+(func1op_call_expression function: _ @function.builtin)
```
Was there any specific reason for this weird syntax node? Or is it having separated nodes for these leads to more efficient parsing?

ObserverOfTime replied:
I assume that's related to [this](https://www.perlmonks.org/?node_id=663393) quirk of Perl.

> ```perl
> whatever  / 25 ; # / ; die "this dies!";
> ```
>
> Schwartz's Snippet can parse two different ways: if `whatever` is nullary (that is, takes no arguments), the first statement is a division in void context, and the rest of the line is a comment. If `whatever` takes an argument, Schwartz's Snippet parses as a call to the `whatever` function with the result of a match operator, then a call to the `die()` function.

leonerd replied:
> Was there any specific reason for this weird syntax node? Or is it having separated nodes for these leads to more efficient parsing?

Short answer: Because perl grammar is crazy.

Longer answer: Perl's parser (actually the lexer - yes really the lexer) categorizes the names of known functions into three separate sets - nullary functions, unary functions, and anything else. The parsing rules for these different categories take account of zero, at-most one or any number of arguments. Because of this, parsing of functions known to be in various categories has different precedence even without parens:

```perl
join $x, $y, $z   # is a regular list function, parses like  join($x, $y, $x)
int $x, $y, $z   # int is known to be a unary function, so this parses like  int($x), $y, $z
```

These rules don't matter for purely syntax-highlighting reasons, so previous technologies like `syntax/perl.vim` didn't have to care. But tooling that attempts to use the tree-sitter node tree result as some sort of AST and try to perform real code analysis will need to know about it, so it makes sense to attmept it here.

lucario387 replied:
TIL more about perl

ObserverOfTime commented on queries/pod/highlights.scm:
```diff
@@ -0,0 +1,40 @@
+
+(verbatim_paragraph (content) @verbatim)
+
+(interior_sequence) @variable
```
```suggestion
(interior_sequence
  (sequence_letter) @character
  (content) @text)
```
(maybe)

rabbiveesh replied:
I left it alone for now b/c it didn't play too nicely with the other interiror seqs below it

ObserverOfTime reviewed (changes requested):
You should also highlight the angle brackets as `@punctutation.delimiter` if possible.

I would suggest adding these as well:
```scheme
(interior_sequence
  (sequence_letter) @character
  (#eq? @letter "X")
  (content) @text.reference)
(interior_sequence
  (sequence_letter) @character
  (#eq? @letter "E")
  (content) @string.escape)
```

clason commented on queries/pod/highlights.scm:
```diff
@@ -0,0 +1,40 @@
+
+(interior_sequence
+  (sequence_letter) @letter
+  (#match? @letter "B")
```
```suggestion
  (#match? @letter "B")
```
prefer `#eq?`, `#any-of?`, and `#lua-match?` (in that order)

ObserverOfTime replied:
I assume `#eq?` can't be used because it might have multiple letters. But `#lua-match?` can.

leonerd replied:
`#eq?` would be fine, I think I just wrote this very early before I realised that `#eq?` even existed; having only seen `#match?` elsewhere

clason commented:
@rabbiveesh we've just switched the format of injection queries to that of upstream, so please rebase and adapt yours. (Holler if you need help, but should be straightforward.)

leonerd commented:
Damn, I literally just yesterday implemented injections in `Text::Treesitter`. Looks like I'm gonna have to alter it again then.

lucario387 commented:
There isn't much to alter, just that only `@injection.language` and `@injection.content` now

clason commented on queries/perl/injections.scm:
```diff
@@ -1 +1,3 @@
-(comments) @comment
+; an injections.scm file for nvim-treesitter
+(comment) @comment
+(pod) @pod
```
```suggestion
((comment) @injection.content
 (#set! injection.language "comment"))
 
((pod) @injection.content
 (#set! injection.language "pod"))
```

clason commented:
> Damn, I literally just yesterday implemented injections in `Text::Treesitter`. Looks like I'm gonna have to alter it again then.

Oh, you were basing your own implementation on Neovim's custom format? That's not a good idea... The new format is the same that tree-sitter (and everybody else) is using, so you should, too ;)

clason commented on queries/pod/highlights.scm:
```diff
@@ -0,0 +1,40 @@
+ (encoding_directive)
+ (cut_directive)] @keyword
+
+(head_paragraph (content) @string)
```
and similar below

rabbiveesh commented:
I'm working on the operator + punctuation highlighting RN, and I'm having an issue. We handle parsing semicolons in our scanner, and I can't for the life of me get tree-sitter to match the semicolon token. Any ideas? 
the obvious
`";" @punctuation.delimiter` is not working, b/c it doesn't come out as a token. And I can't get a match using `#eq?` to work either.

Worst comes to worst i can always rework the grammar to parse semicolon as a normal token, but looking fo rideas

amaanq commented:
alias $._PERLY_SEMICOLON to ';' wherever it's used

ObserverOfTime commented:
Or change the name so it isn't hidden.

rabbiveesh commented:
Rebased + force-pushed so CI can run.

Still open question about the `variable.scalar` and such highlights that i've moved to the bottom of the perl highlights file. Are those not allowed?

ObserverOfTime commented:
> Still open question about the `variable.scalar` and such highlights that i've moved to the bottom of the perl highlights file. Are those not allowed?

They are, but they won't be highlighted any differently by default.

rabbiveesh commented:
Okay, now it's ready!

amaanq commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,107 @@
-   "push" "pop" "shift" "unshift" "splice"
-   "exists" "delete" "keys" "values"
-   "each"))
+(number) @number
+(version) @number
```
```suggestion
[
  (number)
  (version)
] @number
```

amaanq commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,107 @@
-  (grep)
-  (bless)
-] @function.builtin
+(string_literal) @string
+(interpolated_string_literal) @string
+(quoted_word_list) @string
+(command_string) @string
```
```suggestion
[
  (string_literal)
  (interpolated_string_literal)
  (quoted_word_list)
  (command_string)
] @string
```

ObserverOfTime replied:
`(heredoc_content)` can also be grouped here.

amaanq commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,107 @@
+(command_string) @string
 
-;; ----------
+[(heredoc_token) (command_heredoc_token)] @label
+(heredoc_content) @string
+(heredoc_end) @label
```
```suggestion
[
  (heredoc_token)
  (command_heredoc_token)
  (heredoc_end)
] @label
(heredoc_content) @string
```

amaanq commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,107 @@
+(autoquoted_bareword _?) @string.special
 
-(function) @function
+(hash_element_expression key: (bareword) @string.special)
```
hm, would be good to see an example of this, I'm not sure about this one

rabbiveesh replied:
Truth is, i think that never happens, b/c anything that's plausibly a bareword would end up being `autoquoted_bareword` already. I'll remove this rule

rabbiveesh replied:
you can see the normal autoquoty behavior with `$hash->{key}`, `key` there is an autoquoted_bareword

amaanq commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,107 @@
-  (type_glob)
-  (hash_access_variable)
-] @operator
+(goto_expression (label) @label)
+(loopex_expression (label) @label)
```
Maybe just mark any label as a label?
```suggestion
(label) @label
```

amaanq commented:
I'm also a bit concerned about state count, a 9000 large state count might be very laggy/slow on lower end machines, just something to consider

leonerd commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,107 @@
-] @variable
+  "package"
+  "do"
+  "my" "our" "local"
```
It isn't a qualifier, though. A `local` expression temporarily modifies an existing variable, within a scope. It doesn't create a new variable. (It's badly named but we can't change that now ;) )

amaanq replied:
yeah that is badly named, sorry I don't know Perl :)

leonerd commented:
> I'm also a bit concerned about state count, a 9000 large state count might be very laggy/slow on lower end machines, just something to consider

Yeah; we're not happy about that either. A big part of that is because tree-sitter itself doesn't have a native concept of list-associative, or non-associative operators. When we tried to properly implement things like chaining comparisons `1 <= $x <= 10` or the non-associative `isa` operator, these all made the resulting parser.c file enormous, and very slow to generate and compile. We'd love to find a better way to implement them without some of the hackery we have, but we've bounced back and forward between lots of other tree-sitter related folks and not found a solution. I suspect to do it better needs tree-sitter *itself* to gain support for list-associative operators. It was rather beyond our scope to make such a feature addition.

But if you have any way on the process, or even know the right folks to talk to, please do get in touch :) We want that fixed.

ObserverOfTime commented on queries/pod/highlights.scm:
```diff
@@ -0,0 +1,52 @@
+
+(head_paragraph (content) @text.title)
+
+(over_paragraph (content) @text.title)
+(item_paragraph (content) @text.title)
+(encoding_paragraph (content) @text.title)
```
After looking into it a bit more, I'd go with this:

```suggestion
(over_paragraph (content) @number)
(item_paragraph (content) @text)
(encoding_paragraph (content) @string.special)
```

ObserverOfTime commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,107 @@
-  (end_delimiter)
-  (ellipsis_statement)
-] @punctuation.delimiter
+(array_element_expression [array:(_) "->" "[" "]"] @variable.array)
+(slice_expression [array:(_) "->" "[" "]"] @variable.array)
+(keyval_expression [array:(_) "->" "[" "]"] @variable.array)
```
I would rather keep the punctuation highlighted as such.
```suggestion
(array_element_expression array: _ @variable.array)
(slice_expression array: _  @variable.array)
(keyval_expression array: _ @variable.array)
```

ObserverOfTime commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,107 @@
+(keyval_expression [array:(_) "->" "[" "]"] @variable.array)
 
-(function_attribute) @field
+(hash_element_expression [hash:(_) "->" "{" "}"] @variable.hash)
+(slice_expression [hash:(_) "->" "[" "]"] @variable.hash)
+(keyval_expression [hash:(_) "->" "[" "]"] @variable.hash)
```
Same as above.
```suggestion
(array_element_expression hash: _ @variable.hash)
(slice_expression hash: _  @variable.hash)
(keyval_expression hash: _ @variable.hash)
```

ObserverOfTime commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,107 @@
 
-(escape_sequence) @string.escape
+; these are the ones nvim doesn't like
+[(scalar) (arraylen)] @variable.scalar
+(scalar_deref_expression ["->" "$" "*"] @variable.scalar)
+(array) @variable.array
+(array_deref_expression ["->" "@" "*"] @variable.array)
+(hash) @variable.hash
+(hash_deref_expression ["->" "%" "*"] @variable.hash)
```
The "->" should be removed from here so it's highlighted as punctuation.
The rest may or may not be kept as is.

rabbiveesh replied:
I think i'll make it do both `@variable` and also `@variable.$type`, or is that not necessary?

ObserverOfTime replied:
It will fall back to `@variable` anyway.

ObserverOfTime reviewed (changes requested):
Perl is still missing `@punctuation.bracket` highlights, `$#` (maybe `@operator` ?), and regexp modifiers (`@character.special`). Built-in variables (e.g. `ARGV`) should be highlighted as `@variable.builtin` (or `@constant.builtin` where applicable).

Pod should include injections for `=begin` and `=for` blocks. You could also add folds but they aren't necessary.

ObserverOfTime commented:
Also, if you can make the regexp content a named node or field so it can be captured in textobjects, that'd be great.

rabbiveesh commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,107 @@
-  (array_variable)
-  (hash_variable)
-] @variable
+  "package"
```
from what I just read about go it's similar, yes; it sets up the namespace for the following (or containing, depending on if it's block form)

rabbiveesh commented:
> Perl is still missing `@punctuation.bracket` highlights, `$#` (maybe `@operator` ?), some operators (e.g. `&&`), and regexp modifiers (`@character.special`). Built-in variables (e.g. `ARGV`) should be highlighted as `@variable.builtin` (or `@constant.builtin` where applicable).
> 
> Pod should include injections for `=begin` and `=for` blocks. You could also add folds but they aren't necessary.

The comments that I've marked as resolved are a WIP at tree-sitter-perl/tree-sitter-perl#114 . Would like a second pair of eyes there, b/c in my latest commit, I had to alias a literal token to itself. I feel like either me or tree-sitter has done something horribly, horribly wrong, and I'm not sure which; i'd like to get to the bottom of it

ObserverOfTime commented:
> Would like a second pair of eyes there, b/c in my latest commit, I had to alias a literal token to itself.

If I'm not mistaken, literals directly behind hidden nodes cannot be targeted (or inlined).

rabbiveesh commented:
> Perl is still missing ... `$#` (maybe `@operator` ?), 

I'm on the fence if `$#` should be `operator` or `variable`. Syntactically it's a sigil, and even though it gives you information about the variable, i'm not sure if that makes it more of an operator than the `@` in scalar context (`my $len = @ary`) is.

I'm leaning more towards highlighting it as variable, which if I understand the defaulting behavior of the highlighting system, it already should do (b/c we have it captured as `variable.scalar`, unless of course it's suffering from a hidden token bug too)

rabbiveesh commented:
In the following `${name_of_var}`, should those brackets get `@variable` or `@punctuation.bracket`?

Additionally, things like `${_}` technically match a builtin var, should I try and get it to highlight as such or leave that?

ObserverOfTime commented:
> I'm on the fence if `$#` should be `operator` or `variable`.

Well, Lua uses `#` for length (`#foo`) and it's highlighted as  `@operator`.

> In the following `${name_of_var}`, should those brackets get `@variable` or `@punctuation.bracket`?

We typically use `@punctuation.special` for interpolation brackets.

rabbiveesh commented:
> > I'm on the fence if `$#` should be `operator` or `variable`.
> 
> Well, Lua uses `#` for length (`#foo`) and it's highlighted as `@operator`.
> 

Lua doesn't have sigils tho. I'm concerned that the coloring will be odd-ish b/c it'll be the only time that the highlight for the name of the var + the sigil don't match. Happy to do whatever you conclude so long as you're considering that particular point.
Picture for visualization (excuse the poor choice of color) 
![operator-arylen](https://github.com/nvim-treesitter/nvim-treesitter/assets/23664554/5064a0a5-ec71-4e98-bba7-a00f2d658275)


> > In the following `${name_of_var}`, should those brackets get `@variable` or `@punctuation.bracket`?
> 
> We typically use `@punctuation.special` for interpolation brackets.

Ah, well that's a different solution. Let's go for that, then

ObserverOfTime commented:
I'm fine with using `@punctuation.special` for either all or none of the sigils.

amaanq commented:
Status on this? Does upstream still need some changes @rabbiveesh?

rabbiveesh commented:
Ah, sorry, I've been adding a few more features. I should be able to get
back t this tonight.
The POD parser doesn't actually support for blocks yet, so I wanna add that
in, and then we should be good.
I'm working on the state count also, if I forgo list associative parsing
then state count goes down to 1k, so mainly I wish upstream supported it,
but I may have a way to do it with GLR that's less taxing re the state
count

On Tue, Sep 19, 2023, 13:00 Amaan Qureshi ***@***.***> wrote:

> Status on this? Does upstream still need some changes @rabbiveesh
> <https://github.com/rabbiveesh>?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nvim-treesitter/nvim-treesitter/pull/5222#issuecomment-1725195675>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFURPKUUXGRRCI7Y3BKUV53X3FUK7ANCNFSM6AAAAAA3L7UE2Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rabbiveesh commented:
Sorry for the delay; here's the final final PR!

We did not do injections for POD yet, though. @leonerd says that the language names aren't standardized, so we're not sure how much usefulness it'll provide. If this is something that's important for `html`, say, then we'd have to update the parser, but I don't see that as a blocker on this PR as is.

The state count is down by an order of magnitutde b/c I finally gave up on getting listassoc to work, so that's a plus too.

There are some consistency questions that I have, so i'm happy for any + all input on this matter!

clason commented:
The tests already enforce some consistency, so looking at the failures would be a start ;) (You can run them locally with `./script/check-queries.lua`.)

rabbiveesh commented:
I actually meant INTERNAL consistency, not here.

I need to test a different part of the capture in one of the injections, and that's tripping the queries test. How can I do that w/o getting an error (basically, we inject perl into itself when the replacement of a regexp is a perl expression, ie when the modifiers on your `s` operator is a single `e`)

clason commented:
If you only need a capture to refer to it in a predicate, you need to mark it as private with an underscore: `@_modifiers`. _That_ is tripping the test.

rabbiveesh commented:
Ah, thanks a mil, fixed + repushed

amaanq commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,153 @@
-  "{" "}"
-  (standard_input_to_variable)
-] @punctuation.bracket
+(transliteration_content) @string
```
group w/ other strings together too?

amaanq commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,153 @@
-  (regex_option_for_substitution)
-  (regex_option_for_transliteration)
-] @parameter
+(subroutine_declaration_statement name: (_) @function)
```
can a name be different rules? e.g. in Lua it could be like `function foo()` or `function M.foo()`, but you wouldn't want to capture `M.foo`, just `foo`, so ideally capturing identifiers are what you want, but maybe this is correct, not familiar w/ Perl

amaanq commented on queries/perl/injections.scm:
```diff
@@ -1,2 +1,17 @@
+((pod) @injection.content
+ (#set! injection.language "pod"))
+
+(
+  (substitution_regexp
+    (replacement) @injection.content
+    (substitution_regexp_modifiers) @_modifiers
+    ; match if there's a single `e` in the modifiers list
+    (#match? @_modifiers "e")
+    (#not-match? @_modifiers "e.*e")
+  )
+  (#set! injection.language "perl")
+)
```
keep formatting of parenthesis consistent, on the same line is preferred

clason replied:
Apropos: how is the query formatter going? ;)

amaanq replied:
Coming along..just getting really caught up w/ school n' stuff, hope to be done by December :) also it will be upstreamed, and I'm thinking of adding to the cli w/ `tree-sitter query format <path>`

amaanq commented on queries/pod/highlights.scm:
```diff
@@ -0,0 +1,67 @@
+  (sequence_letter) @character
+  (#eq? @character "B")
+  (content) @text.strong)
+(interior_sequence
+  (sequence_letter) @character
+  (#eq? @character "C")
+  (content) @text.literal)
+(interior_sequence
+  (sequence_letter) @character
+  (#eq? @character "F")
+  (content) @text.underline @string.special)
```
add spacing in between each one? gets a bit hard to find start/end w/ so many similar queries

amaanq commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,153 @@
-(function_signature) @type
+(ERROR) @error
+
+(_
+  "{" @punctuation.special
+  (varname)
+  "}" @punctuation.special
+)
+(varname 
+  (block
+    "{" @punctuation.special 
+    "}" @punctuation.special 
+  )
+)
+
+
+(
+  [(varname) (filehandle)] @variable.builtin
+  (#match? @variable.builtin "^((ENV|ARGV|INC|ARGVOUT|SIG|STDIN|STDOUT|STDERR)|[_ab]|\\W|\\d+|\\^.*)$")
+)
```
same, formatting (keep on same line, don't dangle parenthesis)

rabbiveesh commented:
> > the language names aren't standardized, so we're not sure how much usefulness it'll provide
> 
> This is a problem that all markup languages share (#812). I don't know how useful language injections would be in pod's case, but if you can do it _without having to update the parser_ then why not?

well, only `for` blocks would work out of the box. Otherwise, we'd need to add parsing for `begin/end` sections, and you'll have to get @leonerd to do that b/c I tried and it didn't go well. But like i said, we can add it in later + take care of it then

ObserverOfTime commented:
That's fine then.

ObserverOfTime commented on queries/pod/highlights.scm:
```diff
@@ -0,0 +1,67 @@
+
+(command_paragraph
+  (command) @keyword
+  (#match? @keyword "^=head")
```
You should use `lua-match` for these.

rabbiveesh replied:
is there a particular reason? I'd just rather be able to have the same captures available for when i test using the CLI

ObserverOfTime replied:
Performance.

clason replied:
_significant_ performance. I am afraid we hve to insist on that. 

(Queries are editor specific; I understand wanting to minimize the divergence, but it's not possible in general to keep them completely agnostic.)

ObserverOfTime commented on queries/pod/highlights.scm:
```diff
@@ -0,0 +1,67 @@
+
+(interior_sequence
+  (sequence_letter) @character
+  "<" @punctuation.delimiter
+  ">" @punctuation.delimiter
+)
```
```suggestion
  [ "<" ">" ] @punctuation.delimiter)
```

ObserverOfTime commented on queries/perl/injections.scm:
```diff
@@ -1,2 +1,17 @@
+    (replacement) @injection.content
+    (substitution_regexp_modifiers) @_modifiers
+    ; match if there's a single `e` in the modifiers list
+    (#match? @_modifiers "e")
+    (#not-match? @_modifiers "e.*e")
```
```suggestion
    (#lua-match? @_modifiers "e")
    (#not-lua-match? @_modifiers "e.*e")
```

rabbiveesh commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,153 @@
-  (integer)
-  (hexadecimal)
-] @number
+(phaser_statement phase: _ @keyword.phaser)
```
the phaser runs the block at a certain point in the program's lifecycle. for example `BEGIN` runs the code while perl is "compiling" (which is a VERY loose idea, b/c code is run even during "compile" time).
There are also other phasers that exist in the newer class feature, where they are code blocks that run in a specific part of the object's lifecycle. 
I think it makes sense, b/c it is distinct from other kinds of keywords for people that care about it. Looking thru the list of captures, it doesn't seem like there's one that exactly matches

rabbiveesh commented on queries/perl/highlights.scm:
```diff
@@ -1,198 +1,153 @@
 
-(function_definition
-  name: (identifier) @function)
+(_ modifiers: _ @character.special)
```
yes, it's exactly like that example

rabbiveesh commented:
Okay, only question remaining is if i can keep using plain-ol match or if we should switch it to `lua-match`

rabbiveesh commented:
Is CI borked? these failures don't seem to have anything to do w/ my PR

ObserverOfTime commented:
Please avoid merge commits. :)

rabbiveesh commented:
oh sorry, i'm used to a squash merge flow. Let me rebase

clason commented on lockfile.json:
```diff
@@ -414,7 +414,7 @@
   },
   "perl": {
-    "revision": "79e88f64681660f3961939bf764d8f3b4bbb0d27"
+    "revision": "99bbcb0faddfc79db3fd2996cb56689a4685e78c",
```
```suggestion
    "revision": "99bbcb0faddfc79db3fd2996cb56689a4685e78c"
```

This is what's breaking CI.

ObserverOfTime replied:
Gotta love JSON.

rabbiveesh replied:
:facepalm:

clason commented:
> You can squash the commits if you want.

No, please do squash commits, especially if you want to keep perl and pod separate. (In that case, the pod commit needs to be first for the injections to not be broken.)

rabbiveesh commented:
Sure, but could I just squash it down to one? POD doesn't get too much use outside of perl, so it's effectively one commit. Unless you specifically want them separate

clason commented:
That's fine, too.

rabbiveesh commented:
k, squashed; did it in 2 commits b/c the commit log reads bettter that way

clason commented:
@amaanq

amaanq commented:
Thanks for your patience and contribution @rabbiveesh!

rabbiveesh commented:
:partying_face:  thanks for your patience!
